Avoid CPU Sync in SyncBatchNorm When Capturing CUDA Graphs

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #78693
* #78692
* __->__ #78666

We recently updated `SyncBatchNorm` to support empty input batches.
The new code removes stats from ranks with empty inputs. However,
this change breaks CUDA graph capture as it forces CPU sync. This
commit uses `is_current_stream_capturing()` to guard the new code
path, and only run the new code when not capturing CUA Graphs. To
support empty inputs with CUDA graph capturing, we might need to
update CUDA kernels for `batch_norm_backward_elemt` and
`batch_norm_gather_stats_with_counts`. See #78656.

Fixes #78549

Differential Revision: [D36826558](https://our.internmc.facebook.com/intern/diff/D36826558)

mrshenli committed:
Avoid CPU Sync in SyncBatchNorm When Capturing CUDA Graphs

We recently updated `SyncBatchNorm` to support empty input batches.
The new code removes stats from ranks with empty inputs. However,
this change breaks CUDA graph capture as it forces CPU sync. This
commit uses `is_current_stream_capturing()` to guard the new code
path, and only run the new code when not capturing CUA Graphs. To
support empty inputs with CUDA graph capturing, we might need to
update CUDA kernels for `batch_norm_backward_elemt` and
`batch_norm_gather_stats_with_counts`. See #78656.

Fixes #78549

[ghstack-poisoned]

mrshenli commented:
@mrshenli has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36826558).

albanD commented on torch/nn/modules/_functions.py:
```diff
@@ -67,11 +67,19 @@ def forward(self, input, weight, bias, running_mean, running_var, eps, momentum,
+            # FIXME: https://github.com/pytorch/pytorch/issues/78656 describes
+            # a better longer-term solution.
+
+            # remove stats from empty inputs
```
Does that mean that when capturing, this will return wrong results?
Also not sure why is the `count_all[mask]` actually forcing a sync?

mrshenli replied:
> Does that mean that when capturing, this will return wrong results?

Yes. We can change that to if capturing and if `input.numel()`, error out if that's better. 

> Also not sure why is the count_all[mask] actually forcing a sync?

Because the shape of the return value depends on the values in mask?
<img width="702" alt="Screen Shot 2022-06-01 at 5 08 14 PM" src="https://user-images.githubusercontent.com/16999635/171502106-74e3d432-bf8e-441d-8562-23903c72cdb8.png">

albanD replied:
Ho that is a boolean mask. Ok ok!

albanD reviewed (commented):
Should the 1.12 milestone be added to this?

albanD reviewed (approved):
SGTM

mrshenli committed:
Update on "Avoid CPU Sync in SyncBatchNorm When Capturing CUDA Graphs"


We recently updated `SyncBatchNorm` to support empty input batches.
The new code removes stats from ranks with empty inputs. However,
this change breaks CUDA graph capture as it forces CPU sync. This
commit uses `is_current_stream_capturing()` to guard the new code
path, and only run the new code when not capturing CUA Graphs. To
support empty inputs with CUDA graph capturing, we might need to
update CUDA kernels for `batch_norm_backward_elemt` and
`batch_norm_gather_stats_with_counts`. See #78656.

Fixes #78549

Differential Revision: [D36826558](https://our.internmc.facebook.com/intern/diff/D36826558)

[ghstack-poisoned]

mrshenli commented:
@mrshenli has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36826558).

mrshenli committed:
Update on "Avoid CPU Sync in SyncBatchNorm When Capturing CUDA Graphs"


We recently updated `SyncBatchNorm` to support empty input batches.
The new code removes stats from ranks with empty inputs. However,
this change breaks CUDA graph capture as it forces CPU sync. This
commit uses `is_current_stream_capturing()` to guard the new code
path, and only run the new code when not capturing CUA Graphs. To
support empty inputs with CUDA graph capturing, we might need to
update CUDA kernels for `batch_norm_backward_elemt` and
`batch_norm_gather_stats_with_counts`. See #78656.

Fixes #78549

Differential Revision: [D36826558](https://our.internmc.facebook.com/intern/diff/D36826558)

[ghstack-poisoned]

mrshenli commented:
@mrshenli has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36826558).

mrshenli committed:
Update on "Avoid CPU Sync in SyncBatchNorm When Capturing CUDA Graphs"


We recently updated `SyncBatchNorm` to support empty input batches.
The new code removes stats from ranks with empty inputs. However,
this change breaks CUDA graph capture as it forces CPU sync. This
commit uses `is_current_stream_capturing()` to guard the new code
path, and only run the new code when not capturing CUA Graphs. To
support empty inputs with CUDA graph capturing, we might need to
update CUDA kernels for `batch_norm_backward_elemt` and
`batch_norm_gather_stats_with_counts`. See #78656.

Fixes #78549

Differential Revision: [D36826558](https://our.internmc.facebook.com/intern/diff/D36826558)

[ghstack-poisoned]

mrshenli commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -968,6 +969,51 @@ def test_send_recv(self):
+        # capture
+        g = torch.cuda.CUDAGraph()
+        with torch.cuda.graph(g):
+            model(static_input).sum().backward()
```
Hey @mcarilli, I am testing SyncBatchNorm with cuda graph capturing. Looks like allgather is not supported either? But the original code also uses allgather, and I saw your tests have passed allgather. Did I miss anything here?

```
a100-st-p4d24xlarge-7:27858:27858 [0] init.cc:928 NCCL WARN Cuda failure 'operation not permitted when stream is capturing'                                     
a100-st-p4d24xlarge-7:27858:27858 [0] NCCL INFO init.cc:953 -> 1                                                                                                
ERROR:torch.testing._internal.common_distributed:Caught exception:                                                                                              
Traceback (most recent call last):                                                                                                                              
  File "/fsx/users/shenli/pytorch/test/distributed/test_c10d_nccl.py", line 1002, in test_sync_batch_norm_cuda_graph_capture                                    
    model(static_input).sum().backward()                                                                                                                        
  File "/data/home/shenli/.conda/envs/torchdev/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1129, in _call_impl                                
    return forward_call(*input, **kwargs)                                                                                                                       
  File "/data/home/shenli/.conda/envs/torchdev/lib/python3.9/site-packages/torch/nn/modules/container.py", line 139, in forward                                 
    input = module(input)                                                                                                                                       
  File "/data/home/shenli/.conda/envs/torchdev/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1129, in _call_impl                                
    return forward_call(*input, **kwargs)                                                                                                                       
  File "/data/home/shenli/.conda/envs/torchdev/lib/python3.9/site-packages/torch/nn/modules/batchnorm.py", line 748, in forward                                 
    return sync_batch_norm.apply(                                                                                                                               
  File "/data/home/shenli/.conda/envs/torchdev/lib/python3.9/site-packages/torch/nn/modules/_functions.py", line 56, in forward                                 
    dist._all_gather_base(combined_flat, combined, process_group, async_op=False)                                                                               
  File "/data/home/shenli/.conda/envs/torchdev/lib/python3.9/site-packages/torch/distributed/distributed_c10d.py", line 2136, in _all_gather_base               
    work = group._allgather_base(output_tensor, input_tensor)                                                                                                   
RuntimeError: NCCL error in: /fsx/users/shenli/pytorch/torch/csrc/distributed/c10d/NCCLUtils.hpp:126, unhandled cuda error, NCCL version 2.10.3                 
ncclUnhandledCudaError: Call to CUDA function failed.   
```

mrshenli replied:
Do I need to update to the latest nccl 2.12.12-1?

mrshenli replied:
Tested nccl 2.12.12-1, same error

```
1.13.0a0+git94153e1                                                                                            
INFO:torch.distributed.distributed_c10d:Added key: store_based_barrier_key:1 to store for rank: 1                                                               
INFO:torch.distributed.distributed_c10d:Added key: store_based_barrier_key:1 to store for rank: 0             
INFO:torch.distributed.distributed_c10d:Rank 0: Completed store-based barrier for key:store_based_barrier_key:1 with 2 nodes.
INFO:torch.distributed.distributed_c10d:Rank 1: Completed store-based barrier for key:store_based_barrier_key:1 with 2 nodes.                                   
NCCL version 2.12.12+cuda11.1                                                                                                                 
                                                                                                                                                                
a100-st-p4d24xlarge-7:7834:7834 [0] init.cc:938 NCCL WARN Cuda failure 'operation not permitted when stream is capturing'                                       
a100-st-p4d24xlarge-7:7834:7834 [0] NCCL INFO init.cc:963 -> 1                                                               
ERROR:torch.testing._internal.common_distributed:Caught exception:                                                
Traceback (most recent call last):                                                                                                 
  File "/fsx/users/shenli/pytorch/test/distributed/test_c10d_nccl.py", line 1004, in test_sync_batch_norm_cuda_graph_capture                                    
    model(static_input).sum().backward()                                                                                                                        
  File "/fsx/users/shenli/pytorch/torch/nn/modules/module.py", line 1130, in _call_impl                           
    return forward_call(*input, **kwargs)                                                                      
  File "/fsx/users/shenli/pytorch/torch/nn/modules/container.py", line 139, in forward                        
    input = module(input)                                                                                      
  File "/fsx/users/shenli/pytorch/torch/nn/modules/module.py", line 1130, in _call_impl                           
    return forward_call(*input, **kwargs)                                                                         
  File "/fsx/users/shenli/pytorch/torch/nn/modules/batchnorm.py", line 748, in forward                            
    return sync_batch_norm.apply(                                                                                                                               
  File "/fsx/users/shenli/pytorch/torch/nn/modules/_functions.py", line 56, in forward                                             
    dist._all_gather_base(combined_flat, combined, process_group, async_op=False)                                                                                 File "/fsx/users/shenli/pytorch/torch/distributed/distributed_c10d.py", line 2136, in _all_gather_base          
    work = group._allgather_base(output_tensor, input_tensor)                                                     
RuntimeError: NCCL error in: /fsx/users/shenli/pytorch/torch/csrc/distributed/c10d/NCCLUtils.hpp:126, unhandled cuda error, NCCL version 2.12.12
ncclUnhandledCudaError: Call to CUDA function failed.                                                         
Exception raised from create at /fsx/users/shenli/pytorch/torch/csrc/distributed/c10d/NCCLUtils.hpp:126 (most recent call first):                               frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x6b (0x7ff141fc29db in /fsx
/users/shenli/pytorch/torch/lib/libc10.so)      
```

mcarilli replied:
Try
```python
os.environ["NCCL_ASYNC_ERROR_HANDLING"] = "0"
torch.distributed.init_process_group(...)
```
following https://pytorch.org/docs/master/notes/cuda.html#usage-with-distributeddataparallel. That section title really should be "Usage With NCCL"...

mrshenli replied:
Same error, curious, why using `NCCL_ASYNC_ERROR_HANDLING` might make a difference here?

mcarilli replied:
The timeout checking incurs some uncapturable cuda calls (event queries i think, i'd have to double check the code)

mrshenli committed:
Update on "Avoid CPU Sync in SyncBatchNorm When Capturing CUDA Graphs"


We recently updated `SyncBatchNorm` to support empty input batches.
The new code removes stats from ranks with empty inputs. However,
this change breaks CUDA graph capture as it forces CPU sync. This
commit uses `is_current_stream_capturing()` to guard the new code
path, and only run the new code when not capturing CUA Graphs. To
support empty inputs with CUDA graph capturing, we might need to
update CUDA kernels for `batch_norm_backward_elemt` and
`batch_norm_gather_stats_with_counts`. See #78656.

Fixes #78549

Differential Revision: [D36826558](https://our.internmc.facebook.com/intern/diff/D36826558)

[ghstack-poisoned]

mrshenli commented:
Thanks @azrael417 and @mcarilli  for sharing a repro. I verified that the repro failed before this commit and passed afterwards. The repro does not work as a unit test though, still working on addressing that.

mrshenli commented:
@pytorchbot merge this please

mrshenli commented:
@pytorchbot merge

mrshenli commented:
@mrshenli has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36826558).

JerryDaHeLian commented:
torch.distributed.DistBackendError: NCCL error in: ../torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1347, unhandled cuda error (run with NCCL_DEBUG=INFO for details), NCCL version 2.18.5
ncclUnhandledCudaError: Call to CUDA function failed.
Last error:
Cuda failure 'invalid argument'
I have encountered the same problem, has anyone solved it?
